clustermesh: add MCS-API user docs (#35114)
cilium-team/cilium: Documentation/network/clustermesh/mcsapi.rst

.. _gs_clustermesh_mcsapi:

*********************************
Multi-Cluster Services API (Beta)
*********************************

.. include:: ../../beta.rst

> This tutorial will guide you to through the support of `Multi-Cluster Services API (MCS-API) <MCS_API_KEP>`_ in Cilium.
> 
> .. _MCS_API_KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md

Reviewer: ```suggestion
This tutorial will guide you to through the support of `Multi-Cluster Services API (MCS-API)`_ in Cilium.

.. _`Multi-Cluster Services API (MCS-API)`: https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md
```

Otherwise, the link does not seem to be correctly taken into account.

Prerequisites
#############

> You need to have a functioning Cluster Mesh setup, please follow the guide
> :ref:`gs_clustermesh` to set it up.

Reviewer: Nit.

```suggestion
You need to have a functioning Cluster Mesh setup, please follow the 
:ref:`gs_clustermesh` guide to set it up.
```

You first need to install the required MCS-API CRDs:

   .. code-block:: shell-session

>       kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/mcs-api/refs/heads/4690877cc89bcfd9302d516ae43515f446d0aecf/config/crd/multicluster.x-k8s.io_serviceexports.yaml

Reviewer: ```suggestion
      kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/mcs-api/4690877cc89bcfd9302d516ae43515f446d0aecf/config/crd/multicluster.x-k8s.io_serviceexports.yaml
```

      kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/mcs-api/4690877cc89bcfd9302d516ae43515f446d0aecf/config/crd/multicluster.x-k8s.io_serviceimports.yaml

To install Cilium with MCS-API support, run:

>    .. code-block:: shell-session

Reviewer: ```suggestion
   .. parsed-literal::
```

The `|CHART_RELEASE|` substitution does not work inside a code-block. Same for the one below.

      helm install cilium |CHART_RELEASE| \\
      --namespace kube-system \\
      --set clustermesh.enableMCSAPISupport=true

> To enable MCS-API API support on an existing Cilium installation, run:

Reviewer: ```suggestion
To enable MCS-API support on an existing Cilium installation, run:
```

   .. code-block:: shell-session

      helm upgrade cilium |CHART_RELEASE| \\
      --namespace kube-system \\
      --reuse-values \\
      --set clustermesh.enableMCSAPISupport=true
      kubectl -n kube-system rollout restart deployment/cilium-operator

>       kubectl -n kube-system rollout restart deployment/clustermesh-apiserver

Reviewer: Nit. This is not strictly necessary because enabling MCS-API adds an extra parameter, which already causes the deployment to be rolled out.

> Also checkout the `EndpointSlice synchronization <endointslicesync>`_ feature

Reviewer: The link seems broken otherwise.

```suggestion
Also checkout the :ref:`EndpointSlice synchronization <endpointslicesync>` feature
```

if you need Headless Services support.

Installing CoreDNS multicluster
-------------------------------

> You also need to install and configure the `multicluster CoreDNS plugin <multicluster_coredns>`_.
> 
> .. _multicluster_coredns: https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md

Reviewer: Same as above:

```suggestion
You also need to install and configure the `multicluster CoreDNS plugin`_.

.. _`multicluster CoreDNS plugin`: https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md
```

The link doesn't look correct though, as it is the one of the KEP.

The multicluster CoreDNS plugin is currently not provided by default in

> CoreDNS so you will have to recompile it yourself.

Reviewer: I'd maybe add a step to clone the CoreDNS repo, just to make super clear the context of the subsequent operations.

> To rebuild the image you need to add the multicluster plugin to plugins.cfg

Reviewer: ```suggestion
To rebuild the image you need to add the multicluster plugin to the ``plugins.cfg``
```

file. The ordering of plugins matters, add it just below kubernetes plugin
that has very similar functionality:

   .. parsed-literal::

      ...
      kubernetes:kubernetes
      multicluster:github.com/coredns/multicluster
      ...

Then you can build your image simply by running ``make`` and then ``docker build .``
with the right docker tag and upload it to your preferred registry.

> Also you will to make sure that CoreDNS is able to read and watch the relevant

Reviewer: ```suggestion
You also need to make sure that CoreDNS is able to read and watch the relevant
```

MCS-API resources (ServiceImports). You can do so by running the following
command on each cluster:

   .. code-block:: shell-session

      kubectl patch clusterrole system:coredns --type=json \
         -p='[{"op":"add","path":"/rules/-","value":{"apiGroups":["multicluster.x-k8s.io"],"resources":["serviceimports"],"verbs":["list","watch"]}}]'

After that you will need to update the CoreDNS's corefile to also enable the
multicluster plugin, for instance by executing the following command on each cluster:

   .. code-block:: shell-session

      kubectl get configmap -n kube-system coredns -o yaml | \
         sed -E -e 's/^(\s*)kubernetes.*cluster\.local.*$/\1multicluster clusterset.local\n&/' | \
         kubectl replace -f-

And you can finally deploy the CoreDNS image you previously built. Doing so will

> also rollout the CoreDNS deployment and activate the corefile change you previously made.

Author: FYI I hope to be able to simplify this process in the near(-ish) future

Reviewer: That would be awesome :rocket:

         name: rebel-base

In all the clusters and for each set of exported Services that have the same name and namespace,
a ServiceImport resource will be automatically created. All the Endpoints from those exported Services
with the same name and namespace will be merged and made globally available.

An exported Service through MCS-API is available by default on the ``<svc>.<ns>.svc.clusterset.local`` domain.
If you have defined any hostname (via a Statefulset for instance) on your pods
each pods would also be available available through the ``<hostname>.<clustername>.<svc>.<ns>.svc.clusterset.local`` domain.

   .. note::

      The ``<clustername>.<svc>.<ns>.svc.clusterset.local`` domain that would allow
      to get all the endpoints of a Service in a specific cluster is not allowed!

      We recommend creating one service per cluster and/or region and exporting it
      accordingly if you do want to have this kind of behavior, for instance
      creating and exporting services ``mysvc-eu``, ``mysvc-us`` instead of only
      one service. For more information checkout the `dedicated section <MCS_API_KEP_clusterdns>`_
      in the MCS-API KEP explaining this behavior.

> .. _MCS_API_KEP_clusterdns: https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md#not-allowing-cluster-specific-targeting-via-dns

Reviewer: This link should be fixed as well.

> The ServiceImport also has a logic to merge different Service properties:

Reviewer: Nit.

```suggestion
The ServiceImport has also a logic to merge different Service properties:
```

- SessionAffinity
- Ports (Union of the different ServiceExports)
- Type (ClusterSetIP/Headless)

> If any conflict arise on any of those properties, the oldest ServiceExport will

Reviewer: ```suggestion
If any conflict arises on any of these properties, the oldest ServiceExport will
```

have precedence to resolve the conflict. This means that you should get a
consistent behavior globally for the same set of exported Services that has

> the same name and namespace. If any conflicts arise you would be able to see
> details about it in the ServiceExport status Condition.

Reviewer: Nit.

```suggestion
the same name and namespace. If any conflict arises, you would be able to see
details about it in the ServiceExport status Conditions.
```

Deploying a Simple Example Service using MCS-API
------------------------------------------------

1. In cluster 1, deploy:

>    .. code-block:: shell-session

Reviewer: Same here, and below, about the need of using `parsed-literal` rather than `code-block`.
